pkg/distributor/distributor_test.go

```diff
@@ -373,6 +375,313 @@ func Test_TruncateLogLines(t *testing.T) {
+						Entries: []logproto.Entry{},
+						Labels:  generateShardLabels(baseLabels, 0).String(),
+						Hash:    generateShardLabels(baseLabels, 0).Hash(),
+					},
+				},
+				{
+					stream: logproto.Stream{
+						Entries: totalEntries[0:1],
+						Labels:  generateShardLabels(baseLabels, 1).String(),
+						Hash:    generateShardLabels(baseLabels, 1).Hash(),
+					},
+				},
+				{
+					stream: logproto.Stream{
+						Entries: []logproto.Entry{},
+						Labels:  generateShardLabels(baseLabels, 2).String(),
+						Hash:    generateShardLabels(baseLabels, 2).Hash(),
+					},
+				},
+				{
+					stream: logproto.Stream{
+						Entries: totalEntries[1:2],
+						Labels:  generateShardLabels(baseLabels, 3).String(),
+						Hash:    generateShardLabels(baseLabels, 3).Hash(),
+					},
+				},
+			},
+		},
+		{
+			name:    "four shards with 1 entry",
+			shards:  4,
+			entries: totalEntries[0:1],
+			wantDerivedStream: []streamTracker{
+				{
+					stream: logproto.Stream{
+						Labels:  generateShardLabels(baseLabels, 0).String(),
+						Hash:    generateShardLabels(baseLabels, 0).Hash(),
+						Entries: []logproto.Entry{},
+					},
+				},
+				{
+					stream: logproto.Stream{
+						Labels:  generateShardLabels(baseLabels, 1).String(),
+						Hash:    generateShardLabels(baseLabels, 1).Hash(),
+						Entries: []logproto.Entry{},
+					},
+				},
+				{
+					stream: logproto.Stream{
+						Labels:  generateShardLabels(baseLabels, 2).String(),
+						Hash:    generateShardLabels(baseLabels, 2).Hash(),
+						Entries: []logproto.Entry{},
+					},
+				},
+				{
+					stream: logproto.Stream{
+						Entries: totalEntries[0:1],
+						Labels:  generateShardLabels(baseLabels, 3).String(),
+						Hash:    generateShardLabels(baseLabels, 3).Hash(),
+					},
+				},
+			},
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			streamSharder := NewStreamSharderMock(tc.shards)
+			baseStream.Entries = tc.entries
+
+			_, derivedStreams := shardStream(baseStream, Config{}, streamSharder, "fake")
+
+			require.Equal(t, tc.wantDerivedStream, derivedStreams)
+		})
+	}
+}
+
+func BenchmarkShardStream(b *testing.B) {
+	stream := logproto.Stream{}
+	labels := "{app='myapp', job='fizzbuzz'}"
+	lbs, err := syntax.ParseLabels(labels)
+	require.NoError(b, err)
+	stream.Hash = lbs.Hash()
+	stream.Labels = lbs.String()
+
+	// helper funcs
+	generateEntries := func(n int) []logproto.Entry {
+		var entries []logproto.Entry
+		for i := 0; i < n; i++ {
+			entries = append(entries, logproto.Entry{
+				Line:      fmt.Sprintf("log line %d", i),
+				Timestamp: time.Now(),
+			})
+		}
+		return entries
+	}
+	allEntries := generateEntries(25000)
+
+	cfg := Config{ShardStreams: ShardStreamsConfig{
+		Debug: false,
+	}}
+	b.Run("high number of entries, low number of shards", func(b *testing.B) {
+		streamSharder := NewStreamSharderMock(1)
+		stream.Entries = allEntries
+		for n := 0; n < b.N; n++ {
+			shardStream(stream, cfg, streamSharder, "fake")
+		}
```

Comment:
```suggestion
		streamSharder := NewStreamSharderMock(1)
		stream.Entries = allEntries
		b.ResetTimer()
		for n := 0; n < b.N; n++ {
			shardStream(stream, cfg, streamSharder, "fake")
		}
```
If the `NewStreamSharderMock()` takes time you want to exclude it from the benchmark.

Reply:
the allocation is outside of the for loop so it isn't really affected. WDYT?